kernel/rtlil.h

```diff
@@ -85,193 +85,197 @@ namespace RTLIL
 
 		// the global id string cache
 
-		static bool destruct_guard_ok; // POD, will be initialized to zero
-		static struct destruct_guard_t {
-			destruct_guard_t() { destruct_guard_ok = true; }
-			~destruct_guard_t() { destruct_guard_ok = false; }
-		} destruct_guard;
-
-		static std::vector<char*> global_id_storage_;
-		static dict<char*, int, hash_cstr_ops> global_id_index_;
-	#ifndef YOSYS_NO_IDS_REFCNT
-		static std::vector<int> global_refcount_storage_;
-		static std::vector<int> global_free_idx_list_;
-	#endif
-
-	#ifdef YOSYS_USE_STICKY_IDS
-		static int last_created_idx_ptr_;
-		static int last_created_idx_[8];
-	#endif
-
-		static inline void xtrace_db_dump()
-		{
-		#ifdef YOSYS_XTRACE_GET_PUT
-			for (int idx = 0; idx < GetSize(global_id_storage_); idx++)
-			{
-				if (global_id_storage_.at(idx) == nullptr)
-					log("#X# DB-DUMP index %d: FREE\n", idx);
-				else
-					log("#X# DB-DUMP index %d: '%s' (ref %d)\n", idx, global_id_storage_.at(idx), global_refcount_storage_.at(idx));
-			}
+		static struct global_cache_t {
+			bool destruct_guard; // POD, will be initialized to zero
```

Comment:
I think accessing `destruct_guard` of a destroyed object isn't legal. Neither is calling a method of a destroyed object. However, if you put `static bool destruct_guard;` outside of `global_cache_t` that would make it accessible until the program dies, and if you keep all methods accessing it `static` calling them would also be fine.

Does that sound reasonable to you?

Reply:
You are correct that this re-introduces the exact piece of UB that our original patch was aimed at fixing. The compiler can reasonably assume that `destruct_guard` is always true because it cannot be false because accessing it after destruction is strictly invalid. That's why the compiler/LTO was able to optimise this out before our PR.

Reply:
Which other UB?

Reply:
Call any and all functions on this class after it's already been destroyed is itself UB. This guard needing to exist in any form is itself indicative of UB in the way yosys tears down it structures and state as it exits.

Reply:
I did offer a solution to this, though:

> and if you keep all methods accessing it static calling them would also be fine.